File: tools/polyglot_extract.py
#!/usr/bin/env python3
"""
Polyglot Image Payload Extractor
=================================
Extracts and decrypts payloads from polyglot images.

EDUCATIONAL/RESEARCH USE ONLY
This tool demonstrates payload extraction for security research.

Author: SWORDIntel
Date: 2025-11-08
"""

import sys
import os
import argparse
import subprocess
import logging
from pathlib import Path

# Import validation utilities
try:
    from validation_utils import (
        validate_file_exists, validate_output_path, validate_xor_keys,
        atomic_write, setup_logging, ValidationError, FileOperationError
    )
except ImportError:
    # Fallback if run standalone
    sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
    from validation_utils import (
        validate_file_exists, validate_output_path, validate_xor_keys,
        atomic_write, setup_logging, ValidationError, FileOperationError
    )


class PolyglotExtractor:
    """Extracts and decrypts payloads from polyglot images"""

    # Image format EOF markers
    EOF_MARKERS = {
        'gif': b'\x3b',
        'jpg': b'\xff\xd9',
        'jpeg': b'\xff\xd9',
        'png': b'\x49\x45\x4e\x44\xae\x42\x60\x82',
    }

    def __init__(self, verbose=False):
        self.verbose = verbose
        self.logger = setup_logging(verbose=verbose, name='polyglot_extract')

    def log(self, message):
        """Print if verbose mode enabled"""
        if self.verbose:
            self.logger.debug(message)
        else:
            self.logger.info(message)

    def xor_decrypt(self, data, key):
        """XOR decryption (same as encryption)"""
        if isinstance(key, str):
            key = bytes.fromhex(key) if all(c in '0123456789abcdefABCDEF' for c in key) else key.encode()

        decrypted = bytearray(len(data))
        key_len = len(key)

        for i in range(len(data)):
            decrypted[i] = data[i] ^ key[i % key_len]

        return bytes(decrypted)

    def multi_layer_decrypt(self, data, keys):
        """Apply multiple XOR layers in reverse order"""
        result = data
        # Reverse order for decryption
        for key in reversed(keys):
            self.log(f"Removing XOR layer with key: {key if isinstance(key, str) else key.hex()}")
            result = self.xor_decrypt(result, key)
        return result

    def detect_format(self, file_path):
        """Detect image format from header"""
        with open(file_path, 'rb') as f:
            header = f.read(16)

        if header.startswith(b'GIF8'):
            return 'gif'
        elif header.startswith(b'\xff\xd8\xff'):
            return 'jpg'
        elif header.startswith(b'\x89PNG'):
            return 'png'
        else:
            # Try by extension
            ext = Path(file_path).suffix.lower().lstrip('.')
            if ext in self.EOF_MARKERS:
                return ext
            return None

    def find_eof(self, data, img_format):
        """Find EOF marker position"""
        eof_marker = self.EOF_MARKERS.get(img_format)
        if not eof_marker:
            return None

        eof_pos = data.rfind(eof_marker)
        if eof_pos == -1:
            return None

        return eof_pos + len(eof_marker)

    def extract_payload(self, image_path, output_path=None, xor_keys=None, force=False):
        """
        Extract encrypted payload from polyglot image

        Args:
            image_path: Path to polyglot image
            output_path: Where to save extracted payload (default: auto-generate)
            xor_keys: List of XOR keys for decryption
            force: Allow overwriting existing output files

        Returns:
            Tuple of (path_to_extracted_payload, payload_type)

        Raises:
            ValidationError: If inputs are invalid
            FileOperationError: If file operations fail
        """
        # Validate inputs
        try:
            image_path = validate_file_exists(image_path, "Polyglot image")
            xor_keys = validate_xor_keys(xor_keys)
        except ValidationError as e:
            self.logger.error(f"Validation failed: {e}")
            raise

        self.log(f"Reading polyglot file: {image_path}")
        try:
            with open(image_path, 'rb') as f:
                data = f.read()
        except (OSError, IOError) as e:
            raise FileOperationError(f"Failed to read polyglot file: {e}")

        # Detect format
        img_format = self.detect_format(image_path)
        if not img_format:
            raise ValueError("Cannot detect image format")

        self.log(f"Detected format: {img_format.upper()}")

        # Find EOF
        eof_pos = self.find_eof(data, img_format)
        if eof_pos is None:
            raise ValueError(f"Cannot find EOF marker for {img_format}")

        # Check for appended data
        if eof_pos >= len(data):
            raise ValueError("No appended data found (file ends at EOF marker)")

        encrypted_payload = data[eof_pos:]
        self.log(f"Found {len(encrypted_payload):,} bytes after EOF at offset {eof_pos:,}")

        if len(encrypted_payload) == 0:
            raise ValueError("No payload data found")

        # Decrypt
        self.log(f"Decrypting with {len(xor_keys)} layer(s)")
        decrypted = self.multi_layer_decrypt(encrypted_payload, xor_keys)

        # Validate decrypted payload
        if not self._validate_decrypted_payload(decrypted):
            self.logger.warning(
                "Decrypted payload validation failed. "
                "It may be corrupted or encrypted with different keys."
            )

        # Auto-generate output path
        if output_path is None:
            base = Path(image_path).stem
            output_path = f"{base}_extracted_payload.bin"

        # Validate output path
        try:
            output_path = validate_output_path(output_path, allow_overwrite=force)
        except ValidationError as e:
            self.logger.error(f"Invalid output path: {e}")
            raise

        self.log(f"Writing decrypted payload: {output_path}")
        try:
            atomic_write(output_path, decrypted, mode='wb')
        except FileOperationError as e:
            self.logger.error(f"Failed to write payload: {e}")
            raise

        # Detect payload type
        payload_type = self.detect_payload_type(decrypted)

        print(f"\n[+] Payload extracted successfully!")
        print(f"    Image: {image_path}")
        print(f"    EOF position: {eof_pos:,} bytes")
        print(f"    Encrypted size: {len(encrypted_payload):,} bytes")
        print(f"    Decrypted size: {len(decrypted):,} bytes")
        print(f"    Output: {output_path}")
        print(f"    Detected type: {payload_type}")

        return output_path, payload_type

    def _validate_decrypted_payload(self, data: bytes) -> bool:
        """
        Validate that decrypted payload appears legitimate

        Args:
            data: Decrypted payload bytes

        Returns:
            True if payload appears valid
        """
        if len(data) == 0:
            return False

        # Check for known file signatures
        signatures = [
            b'\x7fELF',           # ELF binary
            b'MZ',                # PE/EXE binary
            b'#!/',               # Shebang script
            b'PK\x03\x04',        # ZIP archive
            b'\x89PNG',           # PNG image
            b'GIF8',              # GIF image
            b'\xff\xd8\xff',      # JPEG image
            b'<?xml',             # XML file
            b'<html',             # HTML file
            b'{',                 # JSON/JavaScript (maybe)
        ]

        # Check if starts with known signature
        for sig in signatures:
            if data.startswith(sig):
                return True

        # Check if mostly printable (likely text/script)
        sample_size = min(len(data), 1024)
        printable_count = sum(1 for b in data[:sample_size] if 32 <= b <= 126 or b in (9, 10, 13))
        if printable_count / sample_size > 0.7:
            return True

        # Check for high entropy (might be compressed/encrypted)
        # Very low entropy suggests decryption failed (all zeros, repeated bytes)
        if len(set(data[:1024])) < 10:
            return False

        # If we get here, it's probably valid but unknown format
        return True

    def detect_payload_type(self, data):
        """Detect what kind of payload this is"""
        if data.startswith(b'\x7fELF'):
            return "ELF binary (Linux executable)"
        elif data.startswith(b'MZ'):
            return "PE binary (Windows executable)"
        elif data.startswith(b'#!/'):
            return "Shell script"
        elif data.startswith(b'#!') or b'\n' in data[:100]:
            return "Script file"
        elif data.startswith(b'PK\x03\x04'):
            return "ZIP archive"
        else:
            # Check if mostly printable
            printable = sum(32 <= b <= 126 for b in data[:1024])
            if printable / min(len(data), 1024) > 0.8:
                return "Text/script file"
            return "Binary data (unknown)"


def main():
    parser = argparse.ArgumentParser(
        description='Extract payloads from polyglot images',
        formatter_class=argparse.RawDescriptionHelpFormatter,
        epilog="""
Examples:
  # Extract with default KEYPLUG keys
  %(prog)s suspicious.gif

  # Extract with custom keys
  %(prog)s image.jpg -k d3 -k 410d200d

  # Extract and execute (DANGEROUS!)
  %(prog)s payload.gif --execute

  # Auto-detect and try common keys
  %(prog)s unknown.png --brute-force

WARNING: For educational and authorized security research only!
Never execute payloads from untrusted sources!
        """
    )

    parser.add_argument('image', help='Polyglot image file')
    parser.add_argument('-o', '--output', help='Output file for extracted payload')
    parser.add_argument('-k', '--key', dest='keys', action='append',
                       help='XOR key (hex or string, can be used multiple times)')
    parser.add_argument('-x', '--execute', action='store_true',
                       help='Execute payload after extraction (DANGEROUS!)')
    parser.add_argument('--brute-force', action='store_true',
                       help='Try common KEYPLUG keys automatically')
    parser.add_argument('-v', '--verbose', action='store_true',
                       help='Verbose output')
    parser.add_argument('--force', action='store_true',
                       help='Overwrite output file if it exists')

    args = parser.parse_args()

    extractor = PolyglotExtractor(verbose=args.verbose)

    # Brute force mode
    if args.brute_force:
        common_keys = [
            ['9e'],
            ['d3'],
            ['a5'],
            ['9e', '0a61200d'],
            ['d3', '410d200d'],
            ['9e', '0a61200d', '41414141'],
        ]

        print("[*] Trying common KEYPLUG key combinations...")
        for i, keys in enumerate(common_keys, 1):
            try:
                print(f"\n[*] Attempt {i}/{len(common_keys)}: Keys = {keys}")
                output_path, payload_type = extractor.extract_payload(
                    args.image,
                    args.output or f"payload_attempt_{i}.bin",
                    xor_keys=keys
                )

                # Check if it's a valid payload
                if 'ELF' in payload_type or 'PE' in payload_type or 'script' in payload_type.lower():
                    print(f"\n[+] SUCCESS! Valid payload found with keys: {keys}")
                    break
            except Exception as e:
                print(f"    Failed: {e}")
                continue
        else:
            print("\n[!] No valid payload found with common keys")
            return 1
    else:
        # Normal extraction
        try:
            output_path, payload_type = extractor.extract_payload(
                args.image,
                args.output,
                xor_keys=args.keys,
                force=args.force
            )
        except (ValidationError, FileOperationError) as e:
            print(f"[!] Error: {e}", file=sys.stderr)
            return 1
        except Exception as e:
            print(f"[!] Unexpected error: {e}", file=sys.stderr)
            if args.verbose:
                import traceback
                traceback.print_exc()
            return 1

    # Execute if requested
    if args.execute:
        print(f"\n[!] WARNING: About to execute extracted payload!")
        print(f"[!] Type: {payload_type}")
        response = input("[?] Are you sure? (type 'YES' to confirm): ")

        if response == 'YES':
            print(f"[*] Executing: {output_path}")
            try:
                # Make executable
                os.chmod(output_path, 0o755)
                # Execute
                subprocess.run([output_path], check=True)
            except Exception as e:
                print(f"[!] Execution failed: {e}", file=sys.stderr)
                return 1
        else:
            print("[*] Execution cancelled")

    return 0


if __name__ == '__main__':
    sys.exit(main())

File: archive/vx_underground/polyglot_extract.c
/*
 * polyglot_extract.c - Polyglot Image Payload Extractor
 * ======================================================
 *
 * Extracts and decrypts payloads from polyglot image files.
 * Attributed to APT TeamTNT cryptomining campaigns.
 *
 * Technique discovered in:
 * - APT TeamTNT operations (2020-2024)
 * - APT-41 KEYPLUG malware (2019-2024)
 *
 * For security research and VX Underground publication.
 *
 * Author: SWORDIntel
 * Date: 2025-11-08
 * Attribution: APT TeamTNT
 *
 * BUILD:
 *   gcc -O2 -Wall -o polyglot_extract polyglot_extract.c
 *
 * USAGE:
 *   ./polyglot_extract <polyglot_image> <output> [xor_key]
 *
 * EXAMPLE:
 *   ./polyglot_extract infected.gif payload.bin 9e0a61200d
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <unistd.h>

#define VERSION "1.0.0"
#define MAX_FILE_SIZE (100 * 1024 * 1024)
#define DEFAULT_XOR_KEY "9e0a61200d"

// Image format EOF markers
typedef struct {
    const char *ext;
    const uint8_t *marker;
    size_t marker_len;
    const char *description;
} ImageFormat;

static const uint8_t GIF_HEADER[] = {0x47, 0x49, 0x46, 0x38};  // "GIF8"
static const uint8_t JPEG_HEADER[] = {0xff, 0xd8, 0xff};       // JPEG SOI
static const uint8_t PNG_HEADER[] = {0x89, 0x50, 0x4e, 0x47};  // "\x89PNG"

static const uint8_t GIF_EOF[] = {0x3b};
static const uint8_t JPEG_EOF[] = {0xff, 0xd9};
static const uint8_t PNG_EOF[] = {0x49, 0x45, 0x4e, 0x44, 0xae, 0x42, 0x60, 0x82};

static const ImageFormat FORMATS[] = {
    {"GIF",  GIF_EOF,  sizeof(GIF_EOF),  "GIF87a/89a"},
    {"JPEG", JPEG_EOF, sizeof(JPEG_EOF), "JPEG/JFIF"},
    {"PNG",  PNG_EOF,  sizeof(PNG_EOF),  "PNG"},
    {NULL, NULL, 0, NULL}
};

/*
 * Detect image format from header bytes
 */
const ImageFormat* detect_format_from_data(const uint8_t *data, size_t len) {
    if (len < 8) return NULL;

    if (memcmp(data, GIF_HEADER, sizeof(GIF_HEADER)) == 0) {
        return &FORMATS[0];  // GIF
    } else if (memcmp(data, JPEG_HEADER, sizeof(JPEG_HEADER)) == 0) {
        return &FORMATS[1];  // JPEG
    } else if (memcmp(data, PNG_HEADER, sizeof(PNG_HEADER)) == 0) {
        return &FORMATS[2];  // PNG
    }

    return NULL;
}

/*
 * Find last occurrence of EOF marker
 */
ssize_t find_eof_marker(const uint8_t *data, size_t data_len, const ImageFormat *fmt) {
    if (!data || !fmt || data_len < fmt->marker_len) {
        return -1;
    }

    for (ssize_t i = data_len - fmt->marker_len; i >= 0; i--) {
        if (memcmp(data + i, fmt->marker, fmt->marker_len) == 0) {
            return i + fmt->marker_len;
        }
    }

    return -1;
}

/*
 * Convert hex string to bytes
 */
int hex_to_bytes(const char *hex, uint8_t *bytes, size_t max_len) {
    size_t hex_len = strlen(hex);
    if (hex_len % 2 != 0 || hex_len / 2 > max_len) {
        return -1;
    }

    size_t byte_len = hex_len / 2;
    for (size_t i = 0; i < byte_len; i++) {
        if (sscanf(hex + 2*i, "%2hhx", &bytes[i]) != 1) {
            return -1;
        }
    }

    return byte_len;
}

/*
 * XOR decrypt data
 */
void xor_decrypt(uint8_t *data, size_t data_len, const uint8_t *key, size_t key_len) {
    for (size_t i = 0; i < data_len; i++) {
        data[i] ^= key[i % key_len];
    }
}

/*
 * Detect payload type from decrypted data
 */
const char* detect_payload_type(const uint8_t *data, size_t len) {
    if (len < 4) return "Unknown (too small)";

    // ELF binary
    if (len >= 4 && memcmp(data, "\x7f""ELF", 4) == 0) {
        return "ELF binary (Linux executable)";
    }

    // PE binary
    if (len >= 2 && data[0] == 'M' && data[1] == 'Z') {
        return "PE binary (Windows executable)";
    }

    // Shell script
    if (len >= 2 && data[0] == '#' && data[1] == '!') {
        return "Shell script";
    }

    // ZIP archive
    if (len >= 4 && memcmp(data, "PK\x03\x04", 4) == 0) {
        return "ZIP archive";
    }

    // Check if mostly printable (script/text)
    size_t printable = 0;
    size_t check_len = (len < 1024) ? len : 1024;
    for (size_t i = 0; i < check_len; i++) {
        if ((data[i] >= 32 && data[i] <= 126) || data[i] == '\n' || data[i] == '\t') {
            printable++;
        }
    }

    if ((float)printable / check_len > 0.8) {
        return "Text/script file";
    }

    return "Binary data (unknown type)";
}

/*
 * Read entire file
 */
uint8_t* read_file(const char *filename, size_t *out_len) {
    FILE *fp = fopen(filename, "rb");
    if (!fp) {
        perror("fopen");
        return NULL;
    }

    struct stat st;
    if (fstat(fileno(fp), &st) != 0) {
        perror("fstat");
        fclose(fp);
        return NULL;
    }

    size_t file_size = st.st_size;
    if (file_size == 0 || file_size > MAX_FILE_SIZE) {
        fprintf(stderr, "Invalid file size: %zu bytes\n", file_size);
        fclose(fp);
        return NULL;
    }

    uint8_t *data = malloc(file_size);
    if (!data) {
        perror("malloc");
        fclose(fp);
        return NULL;
    }

    size_t read_len = fread(data, 1, file_size, fp);
    fclose(fp);

    if (read_len != file_size) {
        fprintf(stderr, "Read error\n");
        free(data);
        return NULL;
    }

    *out_len = file_size;
    return data;
}

/*
 * Write data to file
 */
int write_file(const char *filename, const uint8_t *data, size_t data_len) {
    FILE *fp = fopen(filename, "wb");
    if (!fp) {
        perror("fopen");
        return -1;
    }

    size_t written = fwrite(data, 1, data_len, fp);
    fclose(fp);

    if (written != data_len) {
        fprintf(stderr, "Write error\n");
        return -1;
    }

    return 0;
}

/*
 * Main extraction function
 */
int extract_payload(const char *image_path, const char *output_path,
                    const char *xor_key_hex) {

    printf("[*] APT TeamTNT Polyglot Extractor v%s\n", VERSION);
    printf("[*] Polyglot image: %s\n", image_path);
    printf("[*] Output: %s\n", output_path);

    // Read polyglot file
    size_t file_len;
    uint8_t *file_data = read_file(image_path, &file_len);
    if (!file_data) {
        return 1;
    }
    printf("[*] File size: %zu bytes\n", file_len);

    // Detect format
    const ImageFormat *fmt = detect_format_from_data(file_data, file_len);
    if (!fmt) {
        fprintf(stderr, "[!] Cannot detect image format\n");
        free(file_data);
        return 1;
    }
    printf("[*] Detected format: %s (%s)\n", fmt->ext, fmt->description);

    // Find EOF marker
    ssize_t eof_pos = find_eof_marker(file_data, file_len, fmt);
    if (eof_pos < 0) {
        fprintf(stderr, "[!] EOF marker not found\n");
        free(file_data);
        return 1;
    }
    printf("[*] EOF marker at offset: 0x%zx (%zd bytes)\n", eof_pos, eof_pos);

    // Check for appended data
    if ((size_t)eof_pos >= file_len) {
        fprintf(stderr, "[!] No appended data found\n");
        free(file_data);
        return 1;
    }

    size_t payload_len = file_len - eof_pos;
    printf("[*] Appended data size: %zu bytes\n", payload_len);

    // Extract appended data
    uint8_t *encrypted_payload = malloc(payload_len);
    if (!encrypted_payload) {
        perror("malloc");
        free(file_data);
        return 1;
    }
    memcpy(encrypted_payload, file_data + eof_pos, payload_len);
    free(file_data);

    // Parse XOR key
    uint8_t xor_key[256];
    int key_len = hex_to_bytes(xor_key_hex, xor_key, sizeof(xor_key));
    if (key_len < 0) {
        fprintf(stderr, "[!] Invalid XOR key\n");
        free(encrypted_payload);
        return 1;
    }
    printf("[*] XOR key: %s (%d bytes)\n", xor_key_hex, key_len);

    // Decrypt
    printf("[*] Decrypting payload...\n");
    xor_decrypt(encrypted_payload, payload_len, xor_key, key_len);

    // Detect payload type
    const char *payload_type = detect_payload_type(encrypted_payload, payload_len);
    printf("[*] Payload type: %s\n", payload_type);

    // Write decrypted payload
    printf("[*] Writing decrypted payload...\n");
    if (write_file(output_path, encrypted_payload, payload_len) != 0) {
        free(encrypted_payload);
        return 1;
    }

    free(encrypted_payload);

    // Success
    printf("\n[+] Payload extracted successfully!\n");
    printf("    Polyglot file: %zu bytes\n", file_len);
    printf("    EOF position: %zd bytes\n", eof_pos);
    printf("    Encrypted size: %zu bytes\n", payload_len);
    printf("    Payload type: %s\n", payload_type);
    printf("    Output: %s\n", output_path);

    return 0;
}

/*
 * Usage information
 */
void usage(const char *progname) {
    printf("APT TeamTNT Polyglot Extractor v%s\n", VERSION);
    printf("\nUSAGE:\n");
    printf("  %s <polyglot_image> <output> [xor_key]\n\n", progname);
    printf("ARGUMENTS:\n");
    printf("  polyglot_image   Polyglot image file to extract from\n");
    printf("  output           Output file for decrypted payload\n");
    printf("  xor_key          XOR key in hex (default: %s)\n\n", DEFAULT_XOR_KEY);
    printf("EXAMPLES:\n");
    printf("  # Extract with default TeamTNT key\n");
    printf("  %s infected.gif payload.bin\n\n", progname);
    printf("  # Extract with custom key\n");
    printf("  %s stego.jpg malware.sh 9e0a61200d\n\n", progname);
    printf("  # Single-byte XOR\n");
    printf("  %s image.png script.sh d3\n\n", progname);
    printf("SUPPORTED FORMATS:\n");
    printf("  Automatically detects: GIF, JPEG, PNG\n\n");
    printf("PAYLOAD TYPES DETECTED:\n");
    printf("  - ELF binaries (Linux executables)\n");
    printf("  - PE binaries (Windows executables)\n");
    printf("  - Shell scripts (#!/bin/sh, etc)\n");
    printf("  - ZIP archives\n");
    printf("  - Text/script files\n");
    printf("  - Unknown binary data\n\n");
    printf("ATTRIBUTION:\n");
    printf("  Technique: APT TeamTNT cryptomining campaigns\n");
    printf("  Also seen: APT-41 KEYPLUG malware\n");
    printf("  Analysis: SWORDIntel (2025)\n\n");
    printf("For VX Underground publication and security research only.\n");
}

/*
 * Main entry point
 */
int main(int argc, char **argv) {
    if (argc < 3 || argc > 4) {
        usage(argv[0]);
        return 1;
    }

    const char *image_path = argv[1];
    const char *output_path = argv[2];
    const char *xor_key = (argc == 4) ? argv[3] : DEFAULT_XOR_KEY;

    return extract_payload(image_path, output_path, xor_key);
}
